ext/geometry/lib.rs

```diff
@@ -0,0 +1,1101 @@
+const INDEX_M23: usize = 6;
+const INDEX_M24: usize = 7;
+const INDEX_M31: usize = 8;
+const INDEX_M32: usize = 9;
+const INDEX_M33: usize = 10;
+const INDEX_M34: usize = 11;
+const INDEX_M41: usize = 12;
+const INDEX_M42: usize = 13;
+const INDEX_M43: usize = 14;
+const INDEX_M44: usize = 15;
+
+#[op2]
+impl Matrix {
+  #[constructor]
+  #[cppgc]
+  pub fn constructor(#[buffer] buffer: &[f64], is_2d: bool) -> Matrix {
+    Matrix {
+      inner: RefCell::new(Matrix4::from_column_slice(buffer)),
+      is_2d: Cell::new(is_2d),
+    }
+  }
+
+  #[cppgc]
+  pub fn clone(&self) -> Matrix {
+    self.clone()
+  }
+
+  #[fast]
+  #[getter]
+  pub fn a(&self) -> f64 {
+    // SAFETY: in-range access
+    unsafe { *self.inner.borrow().get_unchecked(INDEX_A) }
+  }
+
+  #[fast]
+  #[setter]
+  pub fn a(&self, value: f64) {
+    // SAFETY: in-range access
+    unsafe {
+      *self.inner.borrow_mut().get_unchecked_mut(INDEX_A) = value;
+    }
+  }
+
+  #[fast]
+  #[getter]
+  pub fn b(&self) -> f64 {
+    // SAFETY: in-range access
+    unsafe { *self.inner.borrow().get_unchecked(INDEX_B) }
+  }
+
+  #[fast]
+  #[setter]
+  pub fn b(&self, value: f64) {
+    // SAFETY: in-range access
+    unsafe {
+      *self.inner.borrow_mut().get_unchecked_mut(INDEX_B) = value;
+    }
+  }
+
+  #[fast]
+  #[getter]
+  pub fn c(&self) -> f64 {
+    // SAFETY: in-range access
+    unsafe { *self.inner.borrow().get_unchecked(INDEX_C) }
+  }
+
+  #[fast]
+  #[setter]
+  pub fn c(&self, value: f64) {
+    // SAFETY: in-range access
+    unsafe {
+      *self.inner.borrow_mut().get_unchecked_mut(INDEX_C) = value;
+    }
+  }
+
+  #[fast]
+  #[getter]
+  pub fn d(&self) -> f64 {
+    // SAFETY: in-range access
+    unsafe { *self.inner.borrow().get_unchecked(INDEX_D) }
+  }
+
+  #[fast]
+  #[setter]
+  pub fn d(&self, value: f64) {
+    // SAFETY: in-range access
+    unsafe {
+      *self.inner.borrow_mut().get_unchecked_mut(INDEX_D) = value;
+    }
+  }
+
+  #[fast]
+  #[getter]
+  pub fn e(&self) -> f64 {
+    // SAFETY: in-range access
+    unsafe { *self.inner.borrow().get_unchecked(INDEX_E) }
+  }
+
+  #[fast]
+  #[setter]
+  pub fn e(&self, value: f64) {
+    // SAFETY: in-range access
+    unsafe {
+      *self.inner.borrow_mut().get_unchecked_mut(INDEX_E) = value;
+    }
+  }
+
+  #[fast]
+  #[getter]
+  pub fn f(&self) -> f64 {
+    // SAFETY: in-range access
+    unsafe { *self.inner.borrow().get_unchecked(INDEX_F) }
+  }
+
+  #[fast]
+  #[setter]
+  pub fn f(&self, value: f64) {
+    // SAFETY: in-range access
+    unsafe {
+      *self.inner.borrow_mut().get_unchecked_mut(INDEX_F) = value;
+    }
+  }
+
+  #[fast]
+  #[getter]
+  pub fn m11(&self) -> f64 {
+    // SAFETY: in-range access
+    unsafe { *self.inner.borrow().get_unchecked(INDEX_M11) }
+  }
+
+  #[fast]
+  #[setter]
+  pub fn m11(&self, value: f64) {
+    // SAFETY: in-range access
+    unsafe {
+      *self.inner.borrow_mut().get_unchecked_mut(INDEX_M11) = value;
+    }
```

Comment:
I would suggest a normal array assignment.

```rust
pub fn m11(&self, value: f64) {
  self.inner.borrow_mut().data.0[0][0] = value;
}
```

Since the array index is constant, and since Matrix4 internals are fixed-width arrays and the compiler knows the length of the array, the boundary checks disappear in the compiler's optimization. The execution efficiency will be exactly the same as get_unchecked_mut.

Reply:
Thanks for your suggestion! I will simply do the following to remove unsafe:

```rs
self.inner.borrow_mut().m11 = value;
```

Reply:
Unfortunately, the order of subscripts differs between nalgebra and DOMMatrix, so it would be easier to understand the code using unsafe as it is.